File: framework/backend/csharp/src/K1s0.Cache/Patterns/CacheAside.cs
namespace K1s0.Cache.Patterns;

/// <summary>
/// Implements the cache-aside (lazy-loading) pattern.
/// On cache miss, the provided loader function is called and the result is stored in the cache.
/// </summary>
public sealed class CacheAside
{
    private readonly ICacheOperations _cache;
    private readonly TimeSpan? _defaultTtl;

    /// <summary>
    /// Initializes a new instance of the <see cref="CacheAside"/> class.
    /// </summary>
    /// <param name="cache">The cache operations instance.</param>
    /// <param name="defaultTtl">Optional default TTL for cached entries.</param>
    public CacheAside(ICacheOperations cache, TimeSpan? defaultTtl = null)
    {
        _cache = cache ?? throw new ArgumentNullException(nameof(cache));
        _defaultTtl = defaultTtl;
    }

    /// <summary>
    /// Gets the value from cache if available; otherwise loads it using the provided function and caches the result.
    /// </summary>
    /// <param name="key">The cache key.</param>
    /// <param name="loader">A function that loads the value when the cache does not contain it.</param>
    /// <param name="ttl">Optional TTL override for this specific call.</param>
    /// <param name="ct">A cancellation token.</param>
    /// <returns>The cached or freshly loaded value.</returns>
    public async Task<string> GetOrLoadAsync(
        string key,
        Func<Task<string>> loader,
        TimeSpan? ttl = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(loader);

        string? cached = await _cache.GetAsync(key, ct).ConfigureAwait(false);
        if (cached is not null)
        {
            return cached;
        }

        string value = await loader().ConfigureAwait(false);
        await _cache.SetAsync(key, value, ttl ?? _defaultTtl, ct).ConfigureAwait(false);
        return value;
    }
}

File: framework/backend/python/packages/k1s0-cache/src/k1s0_cache/patterns/cache_aside.py
"""Cache-aside (lazy-loading) pattern."""

from __future__ import annotations

from collections.abc import Awaitable, Callable

from k1s0_cache.operations import CacheOperations


class CacheAside:
    """Implements the cache-aside (lazy-loading) pattern.

    On a cache miss the provided loader is called, and its result is stored
    in the cache before being returned.

    Args:
        cache: Cache operations instance.
        default_ttl: Default TTL applied when none is specified per call.
    """

    def __init__(
        self, cache: CacheOperations, default_ttl: int | None = None
    ) -> None:
        self._cache = cache
        self._default_ttl = default_ttl

    async def get_or_load(
        self,
        key: str,
        loader: Callable[[], Awaitable[str]],
        ttl: int | None = None,
    ) -> str:
        """Return the cached value, or load and cache it on a miss.

        Args:
            key: The cache key.
            loader: An async callable that produces the value on a miss.
            ttl: Optional TTL override.

        Returns:
            The cached or freshly-loaded value.
        """
        value = await self._cache.get(key)
        if value is not None:
            return value

        value = await loader()
        effective_ttl = ttl if ttl is not None else self._default_ttl
        await self._cache.set(key, value, ttl=effective_ttl)
        return value
